### PR TITLE
Add support for anonymous logins

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Primarily, the PR adds support for anonymous logins. That is, we can now declare facades to be used when someone logs in without an auth tag or any credentials. The first such facade is the new cross model relations facade since we will be using macaroons in the api parameter structs when making rpc calls.

A bunch of drive by refactoring was done - login was split up into separate authenticate and other help functions to make it easier to read. Old admin v3 code was consolidated. Various tests needed to be tweaked because they logged in without supplying any macaroons, which now means an anonymous login rather than causing a discharge error.

## QA steps

Run up a cross model relations scenario and verify it all works.

### CODE REVIEW

**Comment by jameinel** [comment]
Login is done on the unauthed API why not just put the other functions on
there?

John
=:->

**Comment by wallyworld** [comment]
That's essentially what's done? We go through the steps of what login does (even for anonymous logins) so that the standard login response is returned to the caller. This response contains controller api addresses and other useful information. We will also in future poke the observer for anonymous logins.

File: apiserver/admin.go [hunk 1]
```diff
@@ -93,39 +179,71 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+	}
+	return result, nil
+}
 
-	entity, lastConnection, err := a.checkCreds(req, isUser)
+func (a *admin) handleAuthError(req params.LoginRequest, kind string, err error) (controllerLogin bool, _ error) {
+	if err == nil {
+		return false, nil
+	}
 	if err != nil {
```
Comment (axw) [inline apiserver/admin.go]: drop the `if err != nil`, it's implied

File: apiserver/admin.go [hunk 2]
```diff
@@ -55,33 +86,88 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+type authResult struct {
+	anonymousLogin      bool
+	userLogin           bool
+	controllerOnlyLogin bool
+	userInfo            *params.AuthUserInfo
+}
+
+func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
+	result := &authResult{
+		userLogin: true,
```
Comment (axw) [inline apiserver/admin.go]: it's confusing to me to set `userLogin: true` for anonymous logins as well. can we please make them mutually exclusive?

File: apiserver/admin.go [hunk 3]
```diff
@@ -55,33 +86,88 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+
+	loginResult := params.LoginResult{
+		Servers:       params.FromNetworkHostsPorts(hostPorts),
+		ControllerTag: model.ControllerTag().String(),
+		UserInfo:      authResult.userInfo,
+		ServerVersion: jujuversion.Current.String(),
+		PublicDNSName: a.srv.publicDNSName(),
+	}
+
+	if authResult.anonymousLogin {
```
Comment (axw) [inline apiserver/admin.go]: I'm not sure about the interaction of anonymous & controller. It should be possible to do an anonymous and contorller-only login I think? Isn't that the case for "crossmodelrelations"?

I think we want to combine the IsAnonymousFacade and IsControllerFacade filter predicates if both conditions are true?

File: apiserver/admin_test.go [hunk 4]
```diff
@@ -1130,3 +1152,70 @@ func (s *migrationSuite) TestExportingModel(c *gc.C) {
...
 	c.Check(err, gc.ErrorMatches, "model migration in progress")
 }
+
+type loginV3Suite struct {
+	loginSuite
+}
+
+var _ = gc.Suite(&loginV3Suite{})
+
+func (s *loginV3Suite) TestClientLoginToEnvironment(c *gc.C) {
```
Comment (axw) [inline apiserver/admin_test.go]: s/Environment/Model/ ?

File: apiserver/admin_test.go [hunk 5]
```diff
@@ -1130,3 +1152,70 @@ func (s *migrationSuite) TestExportingModel(c *gc.C) {
...
+	apiState, err := api.Open(info, api.DialOpts{})
+	c.Assert(err, jc.ErrorIsNil)
+	defer apiState.Close()
+
+	client := apiState.Client()
+	_, err = client.GetModelConstraints()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *loginV3Suite) TestClientLoginToServer(c *gc.C) {
```
Comment (axw) [inline apiserver/admin_test.go]: s/Server/Controller/ ?

File: apiserver/apiserver.go [hunk 6]
```diff
@@ -117,7 +117,7 @@ type Server struct {
 // LoginValidator functions are used to decide whether login requests
 // are to be allowed. The validator is called before credentials are
 // checked.
-type LoginValidator func(params.LoginRequest) error
+type LoginValidator func(authUser string) error
```
Comment (axw) [inline apiserver/apiserver.go]: why did you make this change? if anything I think it should be a names.Tag

File: apiserver/admin.go (hunk 3 above)
Comment (wallyworld) [inline apiserver/admin.go]: "crossmodelrelations" is a model facade.
I didn't think we would necessarily need anonymous controller access but was easy enough to add.

File: apiserver/admin.go (hunk 2 above)
Comment (wallyworld) [inline apiserver/admin.go]: done

File: apiserver/admin.go (hunk 1 above)
Comment (wallyworld) [inline apiserver/admin.go]: doh, refactoring error

File: apiserver/admin_test.go (hunk 4 above)
Comment (wallyworld) [inline apiserver/admin_test.go]: done

File: apiserver/admin_test.go (hunk 5 above)
Comment (wallyworld) [inline apiserver/admin_test.go]: done

File: apiserver/apiserver.go (hunk 6 above)
Comment (wallyworld) [inline apiserver/apiserver.go]: I wanted to narrow the scope of the arg passed to validate. None of the implementations used anything other than req.AuthTag, so why pass the entire request struct and allow unnecessary coupling. Case in point - see maintenanceInProgress() where they wanted to pass in a user but had to construct a whole request struct. I've changed the func to take a tag.

File: apiserver/admin.go [hunk 7]
```diff
@@ -228,6 +163,149 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+		controllerOnlyLogin: a.root.modelUUID == "",
+	}
+	result.userLogin = !result.anonymousLogin
+
+	// Maybe rate limit non-user auth attempts.
+	machineAgent := false
+	if req.AuthTag != "" {
+		var err error
+		kind, err := names.TagKind(req.AuthTag)
+		machineAgent = kind == names.MachineTagKind
```
Comment (axw) [inline apiserver/admin.go]: move this under the following if?

File: apiserver/admin.go [hunk 8]
```diff
@@ -228,6 +163,149 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
...
+		var err error
+		result.userInfo, err = a.checkUserPermissions(userTag, result.controllerOnlyLogin)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		result.userInfo.LastConnection = lastConnection
+	}
+	if result.controllerOnlyLogin {
+		if result.anonymousLogin {
+			logger.Debugf(" anonymous controller login")
```
Comment (axw) [inline apiserver/admin.go]: what's up with the whitespace? and below

Comment (wallyworld) [inline apiserver/admin.go]: NFI, it's correct  locally

File: apiserver/admin.go (hunk 7 above)
Comment (wallyworld) [inline apiserver/admin.go]: ok

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by manadart** [comment]
[Update juju/description and Migrations for New Space ID](https://trello.com/c/Neik2CkC/1638-update-juju-description-and-migrations-for-new-space-id)

**Pull request merged**